libhwy: 1.0.7 -> 1.3.0, gcc15: backport aarch64 ICE on vreg split

trofi opened the pull request:
Changes:
- https://github.com/google/highway/releases/tag/1.1.0
- https://github.com/google/highway/releases/tag/1.2.0
- https://github.com/google/highway/releases/tag/1.3.0

## Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/4305

SuperSandro2000 commented:
@ofborg build libhwy

SuperSandro2000 commented:
aarch64-linux fails to build

```
FAILED: CMakeFiles/math_test.dir/hwy/contrib/math/math_test.cc.o 
/nix/store/s6g4y3da8f87wwshs3l86wfcycmjgdqi-gcc-wrapper-13.3.0/bin/g++ -DHWY_STATIC_DEFINE -I/build/source -isystem /nix/store/s0ncp3pvk2vzsiv5pi8hb0p35dm1d8ch-gtest-1.14.0-dev/include -O3 -DNDEBUG -std=c++17 -fPIE -fvisibility=hidden -fvisibility-inlines-hidden -Wno-builtin-macro-redefined -D__DATE__=\"redacted\" -D__TIMESTAMP__=\"redacted\" -D__TIME__=\"redacted\" -fmerge-all-constants -Wall -Wextra -Wconversion -Wsign-conversion -Wvla -Wnon-virtual-dtor -Wcast-align -fmath-errno -fno-exceptions -Wno-psabi -DHWY_IS_TEST=1 -MD -MT CMakeFiles/math_test.dir/hwy/contrib/math/math_test.cc.o -MF CMakeFiles/math_test.dir/hwy/contrib/math/math_test.cc.o.d -o CMakeFiles/math_test.dir/hwy/contrib/math/math_test.cc.o -c /build/source/hwy/contrib/math/math_test.cc
In file included from /build/source/hwy/tests/test_util-inl.h:28,
                 from /build/source/hwy/contrib/math/math_test.cc:31,
                 from /build/source/hwy/foreach_target.h:163,
                 from /build/source/hwy/contrib/math/math_test.cc:28:
/build/source/hwy/tests/test_util.h: In function 'std::string hwy::TypeName(T, size_t) [with T = double]':
/build/source/hwy/tests/test_util.h:152:19: error: this operation requires the SVE ISA extension
  152 |   detail::TypeName(detail::MakeTypeInfo<T>(), N, string100);
      |   ~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/build/source/hwy/tests/test_util.h:152:19: note: you can enable SVE using the command-line option '-march', or by using the 'target' attribute or pragma
```

NickCao reviewed (commented):
The riscv patch need an update:

```diff
diff --git a/pkgs/development/libraries/libhwy/default.nix b/pkgs/development/libraries/libhwy/default.nix
index c82d3738efe9..c5a777be8691 100644
--- a/pkgs/development/libraries/libhwy/default.nix
+++ b/pkgs/development/libraries/libhwy/default.nix
@@ -18,14 +18,13 @@ stdenv.mkDerivation rec {
     hash = "sha256-yJQH5ZkpEdJ6lsTAt6yJSN3TQnVoxNpkbChENaxhcHo=";
   };

-  patches = lib.optional stdenv.hostPlatform.isRiscV
-    # Adds CMake option HWY_CMAKE_RVV
-    # https://github.com/google/highway/pull/1743
+  patches = [
     (fetchpatch {
-      name = "libhwy-add-rvv-optout.patch";
-      url = "https://github.com/google/highway/commit/5d58d233fbcec0c6a39df8186a877329147324b3.patch";
-      hash = "sha256-ileSNYddOt1F5rooRB0fXT20WkVlnG+gP5w7qJdBuww=";
-    });
+      name = "disable-RVV-runtime-dispatch.patch";
+      url = "https://github.com/google/highway/commit/c95cc0237d2f7a0f5ca5dc3fb4b5961b2b1dcdfc.patch";
+      hash = "sha256-oQfyZrjZ9MGcSrFInbbj+0iOLjPng7tgTzli1QTITSg=";
+    })
+  ];

   hardeningDisable = lib.optionals stdenv.hostPlatform.isAarch64 [
     # aarch64-specific code gets:
```

trofi commented:
> The riscv patch need an update:

Applied. Thank you!

SuperSandro2000 commented:
> aarch64-linux fails to build

that still doesn't work...

trofi commented:
Yeah, I did not get to look at it. Let's convert to draft.

philiptaron commented:
7ea13320bf8eedced05d0f5545f5b90169c7a100 landed in the interim. Maybe aarch64 works now?

trofi committed:
libhwy: 1.0.7 -> 1.3.0

Changes:
- https://github.com/google/highway/releases/tag/1.1.0
- https://github.com/google/highway/releases/tag/1.2.0
- https://github.com/google/highway/releases/tag/1.3.0

trofi commented:
Sounds good. Let's try again.

trofi commented:
Ofborg says it still fails. Looks like a `-march=` / target attribute mismatch:

```
In file included from /build/source/hwy/tests/test_util-inl.h:28,
                 from /build/source/hwy/contrib/math/math_test.cc:31,
                 from /build/source/hwy/foreach_target.h:163,
                 from /build/source/hwy/contrib/math/math_test.cc:28:
/build/source/hwy/tests/test_util.h: In function 'std::string hwy::TypeName(T, size_t) [with T = float]':
/build/source/hwy/tests/test_util.h:152:19: error: this operation requires the SVE ISA extension
  152 |   detail::TypeName(detail::MakeTypeInfo<T>(), N, string100);
      |   ~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/build/source/hwy/tests/test_util.h:152:19: note: you can enable SVE using the command-line option '-march', or by using the 'target' attribute or pragma
```

trofi commented:
> I know it is stale, but new version released: https://github.com/google/highway/releases/tag/1.3.0

Good catch! Updated to `1.3.0`. Did not test on `aarch64-linux` as I don't have it.

RossComputerGuy commented:
Btw, I figured out this is a GCC bug. https://github.com/google/highway/issues/2250

This fixes it for my Ampere desktop:
```
    ++ lib.optionals (stdenv.hostPlatform.isAarch64 && stdenv.cc.isGNU) [
      # Workaround for a GCC bug https://github.com/google/highway/issues/2250
      (lib.cmakeFeature "CMAKE_CXX_FLAGS_RELEASE" "-O2")
    ]
```

trofi commented:
As `libhwy` is a mostly headers-only library the workaround would probably need to be applied to all broken users, not just `libhwy`.

RossComputerGuy commented:
Maybe, though I managed to compile with only setting it in libhwy. I managed to compile down to netlify-cli. I can try more packages but idk what to build which might be affected.

trofi commented:
There are about 4 `libjxl` users in `nixpkgs`:

```
$ git grep -F 'libhwy,' | cat
pkgs/by-name/gd/gdal/package.nix:  libhwy,
pkgs/by-name/li/libjxl/package.nix:  libhwy,
pkgs/by-name/ss/ssimulacra2/package.nix:  libhwy,
pkgs/by-name/vi/vips/package.nix:  libhwy,
```

On my system it's `gdal` (via `dav1d`) and `libjxl` (via various image readers).

philiptaron commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 320616 --package libhwy`
Commit: `dcc840911fcae408a45c203c9f626f3a7ea40d74`

---
### `aarch64-linux`
<details>
  <summary>:x: 1 package failed to build:</summary>
  <ul>
    <li>libhwy</li>
  </ul>
</details>

---
<details>
<summary>Error logs: `aarch64-linux`</summary>
<details>
<summary>libhwy</summary>
<pre>In file included from /build/source/hwy/tests/test_util-inl.h:31,
                 from /build/source/hwy/contrib/math/math_test.cc:31,
                 from /build/source/hwy/foreach_target.h:174,
                 from /build/source/hwy/contrib/math/math_test.cc:28:
/build/source/hwy/tests/test_util.h: In function &#x27;std::string hwy::TypeName(T, size_t) [with T = float]&#x27;:
/build/source/hwy/tests/test_util.h:182:19: error: this operation requires the SVE ISA extension
  182 |   detail::TypeName(detail::MakeTypeInfo&lt;T&gt;(), N, string100);
      |   ~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/build/source/hwy/tests/test_util.h:182:19: note: you can enable SVE using the command-line option &#x27;-march&#x27;, or by using the &#x27;target&#x27; attribute or pragma
[114/215] Building CXX object CMakeFiles/masked_arithmetic_test.dir/hwy/tests/masked_arithmetic_test.cc.o
[115/215] Building CXX object CMakeFiles/reduction_test.dir/hwy/tests/reduction_test.cc.o
[116/215] Building CXX object CMakeFiles/sort_test.dir/hwy/contrib/sort/sort_test.cc.o
[117/215] Building CXX object CMakeFiles/copy_test.dir/hwy/contrib/algo/copy_test.cc.o
[118/215] Building CXX object CMakeFiles/memory_test.dir/hwy/tests/memory_test.cc.o
[119/215] Building CXX object CMakeFiles/transform_test.dir/hwy/contrib/algo/transform_test.cc.o
[120/215] Building CXX object CMakeFiles/find_test.dir/hwy/contrib/algo/find_test.cc.o
[121/215] Building CXX object CMakeFiles/shift_test.dir/hwy/tests/shift_test.cc.o
[122/215] Building CXX object CMakeFiles/dot_test.dir/hwy/contrib/dot/dot_test.cc.o
[123/215] Building CXX object CMakeFiles/bit_pack_test.dir/hwy/contrib/bit_pack/bit_pack_test.cc.o
ninja: build stopped: subcommand failed.</pre>
</details>
</details>

trofi commented:
> ## `nixpkgs-review` result
> Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).
> 
> Command: `nixpkgs-review pr 320616 --package libhwy` Commit: `dcc840911fcae408a45c203c9f626f3a7ea40d74`
> 
> ### `aarch64-linux`
> ❌ 1 package failed to build:
> * libhwy
> 
> Error logs: `aarch64-linux`
> libhwy
> In file included from /build/source/hwy/tests/test_util-inl.h:31,
>                  from /build/source/hwy/contrib/math/math_test.cc:31,
>                  from /build/source/hwy/foreach_target.h:174,
>                  from /build/source/hwy/contrib/math/math_test.cc:28:
> /build/source/hwy/tests/test_util.h: In function 'std::string hwy::TypeName(T, size_t) [with T = float]':
> /build/source/hwy/tests/test_util.h:182:19: error: this operation requires the SVE ISA extension
>   182 |   detail::TypeName(detail::MakeTypeInfo<T>(), N, string100);
>       |   ~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
> /build/source/hwy/tests/test_util.h:182:19: note: you can enable SVE using the command-line option '-march', or by using the 'target' attribute or pragma
> [114/215] Building CXX object CMakeFiles/masked_arithmetic_test.dir/hwy/tests/masked_arithmetic_test.cc.o
> [115/215] Building CXX object CMakeFiles/reduction_test.dir/hwy/tests/reduction_test.cc.o
> [116/215] Building CXX object CMakeFiles/sort_test.dir/hwy/contrib/sort/sort_test.cc.o
> [117/215] Building CXX object CMakeFiles/copy_test.dir/hwy/contrib/algo/copy_test.cc.o
> [118/215] Building CXX object CMakeFiles/memory_test.dir/hwy/tests/memory_test.cc.o
> [119/215] Building CXX object CMakeFiles/transform_test.dir/hwy/contrib/algo/transform_test.cc.o
> [120/215] Building CXX object CMakeFiles/find_test.dir/hwy/contrib/algo/find_test.cc.o
> [121/215] Building CXX object CMakeFiles/shift_test.dir/hwy/tests/shift_test.cc.o
> [122/215] Building CXX object CMakeFiles/dot_test.dir/hwy/contrib/dot/dot_test.cc.o
> [123/215] Building CXX object CMakeFiles/bit_pack_test.dir/hwy/contrib/bit_pack/bit_pack_test.cc.o
> ninja: build stopped: subcommand failed.

Do you know if it was built from branch head (still on gcc-14), or branch merge (gcc-15)? If it's head I can rebase to get gcc-15.

philiptaron commented:
I just used Nixpkgs-review which I believe uses the hypothetical merge commit by default. I'm game for a rebase though, my spark isn't doing anything else today

philiptaron commented:
Same exact error.

```
$ gcc --version
gcc (GCC) 15.2.0
Copyright (C) 2025 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

trofi commented:
Thank you! Switching back to draft.

RossComputerGuy commented:
> Btw, I figured out this is a GCC bug. [google/highway#2250](https://github.com/google/highway/issues/2250)
> 
> This fixes it for my Ampere desktop:
> 
> ```
>     ++ lib.optionals (stdenv.hostPlatform.isAarch64 && stdenv.cc.isGNU) [
>       # Workaround for a GCC bug https://github.com/google/highway/issues/2250
>       (lib.cmakeFeature "CMAKE_CXX_FLAGS_RELEASE" "-O2")
>     ]
> ```

This has been working for me until recently. Now, an even worse error is occurring:

```
libhwy> FAILED: [code=1] CMakeFiles/random_test.dir/hwy/contrib/random/random_test.cc.o
libhwy> /nix/store/484f8q8xa48icqgdq9b67vhq8clcqjy9-gcc-wrapper-15.2.0/bin/g++ -DHWY_STATIC_DEFINE -I/build/source -isystem /nix/store/3fngz3l2qk48gjdcqqbrnl410wyh82w2-gtest-1.17.0-dev/include -O2 -std=c++17 -fPIE -fvisibility=hidden -fvisibility-inlines-hidden -Wno-builtin-macro-redefined -D__DATE__=\"redacted\" -D__TIMESTAMP__=\"redacted\" -D__TIME__=\"redacted\" -fmerge-all-constants -Wall -Wextra -Wconversion -Wsign-conversion -Wvla -Wnon-virtual-dtor -Wcast-align -fmath-errno -fno-exceptions -Wno-psabi -DHWY_IS_TEST=1 -MD -MT CMakeFiles/random_test.dir/hwy/contrib/random/random_test.cc.o -MF CMakeFiles/random_test.dir/hwy/contrib/random/random_test.cc.o.d -o CMakeFiles/random_test.dir/hwy/contrib/random/random_test.cc.o -c /build/source/hwy/contrib/random/random_test.cc
libhwy> In file included from /build/source/hwy/contrib/random/random_test.cc:19,
libhwy>                  from /build/source/hwy/foreach_target.h:196,
libhwy>                  from /build/source/hwy/contrib/random/random_test.cc:17:
libhwy> /build/source/hwy/contrib/random/random-inl.h: In constructor 'hwy::N_SVE::VectorXoshiro::VectorXoshiro(uint64_t, uint64_t)':
libhwy> /build/source/hwy/contrib/random/random-inl.h:199:3: error: unrecognizable insn:
libhwy>   199 |   }
libhwy>       |   ^
libhwy> (insn 35 7 36 2 (set (reg:TI 163 [ D.337642 ])
libhwy>         (subreg:TI (const_vector:V2DI [
libhwy>                     (const_int 4 [0x4])
libhwy>                     (const_poly_int:DI [2, 2])
libhwy>                 ]) 0)) "/build/source/hwy/contrib/random/random-inl.h":183:9 -1
libhwy>      (nil))
libhwy> during RTL pass: vregs
libhwy> /build/source/hwy/contrib/random/random-inl.h:199:3: internal compiler error: in extract_insn, at recog.cc:2882
libhwy> 0x219bd5b diagnostic_context::diagnostic_impl(rich_location*, diagnostic_metadata const*, diagnostic_option_id, char const*, std::__va_list*, diagnostic_t)
libhwy>         ???:0
libhwy> 0x21aedf7 internal_error(char const*, ...)
libhwy>         ???:0
libhwy> 0x7f154f fancy_abort(char const*, int, char const*)
libhwy>         ???:0
libhwy> 0x7ef277 _fatal_insn(char const*, rtx_def const*, char const*, int, char const*)
libhwy>         ???:0
libhwy> 0x7ef2a3 _fatal_insn_not_found(rtx_def const*, char const*, int, char const*)
libhwy>         ???:0
libhwy> 0xfd618f extract_insn(rtx_insn*)
libhwy>         ???:0
libhwy> 0xce126f (anonymous namespace)::pass_instantiate_virtual_regs::execute(function*)
libhwy>         ???:0
libhwy> Please submit a full bug report, with preprocessed source (by using -freport-bug).
libhwy> Please include the complete backtrace with any bug report.
libhwy> See <https://gcc.gnu.org/bugs/> for instructions.
```

Disabling the tests on aarch64-linux has gotten it to complete. Packages which depend on libhwy continue to function fine.

trofi commented:
> `gcc-16` does not crash it any more. Probably got fixed. I'll try to find a fix.

Was already fixed in `gcc-15` branch (will be a part of upcoming `gcc-15.3.0`) as https://gcc.gnu.org/git/?p=gcc.git;a=commitdiff;h=d755aa03db0ad5b71ee7f39b09c92870789f2f00 (from http://gcc.gnu.org/PR120718 ).

trofi committed:
gcc15: backport aarch64 ICE on vreg split

Without the change `libhwy-1.3.0` fails as:

    libhwy-aarch64-unknown-linux-gnu> /build/source/hwy/contrib/random/random-inl.h: In function 'hwy::N_SVE2::VectorXoshiro::VectorXoshiro(uint64_t, uint64_t)':
    libhwy-aarch64-unknown-linux-gnu> /build/source/hwy/contrib/random/random-inl.h:199:3: error: unrecognizable insn:
    libhwy-aarch64-unknown-linux-gnu>   199 |   }
    libhwy-aarch64-unknown-linux-gnu>       |   ^
    libhwy-aarch64-unknown-linux-gnu> (insn 18 4 19 2 (set (reg:TI 154 [ D.606722 ])
    libhwy-aarch64-unknown-linux-gnu>         (subreg:TI (const_vector:V2DI [
    libhwy-aarch64-unknown-linux-gnu>                     (const_int 4 [0x4])
    libhwy-aarch64-unknown-linux-gnu>                     (const_poly_int:DI [2, 2])
    libhwy-aarch64-unknown-linux-gnu>                 ]) 0)) "/build/source/hwy/contrib/random/random-inl.h":183:9 -1
    libhwy-aarch64-unknown-linux-gnu>      (nil))
    libhwy-aarch64-unknown-linux-gnu> during RTL pass: vregs
    libhwy-aarch64-unknown-linux-gnu> /build/source/hwy/contrib/random/random-inl.h:199:3: internal compiler error: in extract_insn, at recog.cc:2882
    libhwy-aarch64-unknown-linux-gnu> 0x233d5bb diagnostic_context::diagnostic_impl(rich_location*, diagnostic_metadata const*, diagnostic_option_id, char const*, __va_list_tag (*) [1], diagnostic_t)

Co-authored-by: Philip Taron <philip.taron@gmail.com>

philiptaron reviewed (commented):
Doing a build.
